File: lintcode/longest_words/cpp/longest_words.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/longest-words
@Language: C++
@Datetime: 17-02-15 15:52
*/

#include <algorithm>

class Solution {
public:
    /**
     * @param dict: a vector of strings
     * @return: a vector of strings
     */
    std::vector<std::string> longestWords(std::vector<std::string> &dict)
    {
        std::vector<std::string> result;
        
        const auto maxLength = max_element(dict.cbegin(), dict.cend(),
            [] (decltype(*dict.cbegin()) left, decltype(*dict.cbegin()) right) {
                return left.size() < right.size();
            })->size();
            
        for (const auto &str : dict) {
            if (str.size() == maxLength) {
                result.push_back(str);
            }
        }
        
        
        return result;
    }
};

File: lintcode/easy/longest_words/py/longest_words.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/longest-words
@Language: Python
@Datetime: 17-02-15 13:55
'''

class Solution:
    # @param dictionary: a list of strings
    # @return: a list of strings
    def longestWords(self, dictionary):
        ordered = sorted(dictionary, key=len, reverse=True)
        words   = []
        
        for i in range(len(ordered)):
            words.append(ordered[i])
            
            if i + 1 < len(ordered) and len(ordered[i]) != len(ordered[i + 1]):
                break
                
        return words